engine/backends/fstar/fstar_backend.ml

```diff
@@ -152,61 +152,74 @@ struct
-                span
+        Error.unimplemented
+          ~details:
+            "Integers cannot be printed as constants, they can only be printed \
+             as expressions."
+          span
+    | Float _ ->
+        Error.unimplemented
+          ~details:
+            "Floats cannot be printed as constants, they can only be printed \
+             as expressions."
+          span
+    | Bool b -> F.Const.Const_bool b
+
+  (* Print a literal appearing in a pattern as an F* pattern *)
+  let rec pliteral_as_pat span (e : literal) =
+    match e with
+    | Int { value; kind = { size; signedness }; negative } ->
+        let pat_name =
+          F.pat
+          @@ F.AST.PatName (F.lid [ "Rust_primitives"; "Integers"; "MkInt" ])
+        in
+        let mk_const c = F.AST.PatConst c |> F.pat in
+        let mk_int value negative =
+          mk_const (F.Const.Const_int (pnegative negative ^ value, None))
         in
-        F.Const.Const_int
-          ( pnegative negative ^ value,
-            Some
-              ( (match signedness with Signed -> Signed | Unsigned -> Unsigned),
-                size ) )
+        F.pat_app pat_name @@ [ mk_int value negative ]
     | Float _ ->
         Error.unimplemented ~issue_id:464
-          ~details:"Matching on f32 or f64 literals is not yet supported." span
-    | Bool b -> F.Const.Const_bool b
+          ~details:"Pattern matching on floats is not yet supported." span
+    | _ -> F.pat @@ F.AST.PatConst (pliteral_as_const span e)
 
   let pliteral_as_expr span (e : literal) =
     let mk_const c = F.AST.Const c |> F.term in
-    let mk_nat value negative =
+    let mk_int value negative =
       mk_const (F.Const.Const_int (pnegative negative ^ value, None))
     in
-    let wrap_app fn x n = F.mk_e_app (F.term_of_lid [ fn ]) [ mk_nat x n ] in
     match e with
-    | Int { value; kind = { size = SSize; signedness = sn }; negative = n } ->
-        wrap_app (match sn with Signed -> "isz" | Unsigned -> "sz") value n
-    | Int { value; kind = { size = S128; signedness = sn }; negative } ->
-        let prefix = match sn with Signed -> "i" | Unsigned -> "u" in
-        wrap_app ("pub_" ^ prefix ^ "128") value negative
+    | Int { value; kind = { size; signedness }; negative = n } ->
+        let f =
+          match (size, signedness) with
+          | S8, Signed -> F.lid [ "mk_i8" ]
+          | S16, Signed -> F.lid [ "mk_i16" ]
+          | S32, Signed -> F.lid [ "mk_i32" ]
+          | S64, Signed -> F.lid [ "mk_i64" ]
+          | S128, Signed -> F.lid [ "mk_i128" ]
+          | SSize, Signed -> F.lid [ "mk_isize" ]
+          | S8, Unsigned -> F.lid [ "mk_u8" ]
+          | S16, Unsigned -> F.lid [ "mk_u16" ]
+          | S32, Unsigned -> F.lid [ "mk_u32" ]
+          | S64, Unsigned -> F.lid [ "mk_u64" ]
+          | S128, Unsigned -> F.lid [ "mk_u128" ]
+          | SSize, Unsigned -> F.lid [ "mk_usize" ]
+        in
+        F.mk_e_app (F.term @@ F.AST.Name f) [ mk_int value n ]
```

Comment:
I think the `pliteral_as_expr` / `pliteral_as_pat` no longer makes sense, we should just have a `pliteral` with a flag ``mode:[`Expr | `Pat]``.
(now we have runtime errors that we should kill statically)

If you want I can implement it, or we can also decide to ignore that, since we want to drop that backend (but I don't see that happening too soon)

Reply:
Let's ignore that for now and come back to this as part of a more general backend rewriting task.